lighthouse-core/lib/lantern-trace-saver.js

```diff
@@ -0,0 +1,190 @@
+
+  // Create a fake requestId counter
+  let requestId = 1;
+  let lastEventEndTime = 0;
+  for (const [node, timing] of nodeTimings.entries()) {
+    lastEventEndTime = Math.max(lastEventEndTime, timing.endTime);
+    if (node.type === 'cpu') {
+      // Represent all CPU work that was bundled in a task as an EvaluateScript event
+      const cpuNode = /** @type {LH.Gatherer.Simulation.GraphCPUNode} */ (node);
+      traceEvents.push(createFakeTaskEvent(cpuNode, timing));
+    } else {
+      const networkNode = /** @type {LH.Gatherer.Simulation.GraphNetworkNode} */ (node);
+      // Ignore data URIs as they don't really add much value
+      if (/^data/.test(networkNode.record.url)) continue;
+      traceEvents.push(...createFakeNetworkEvents(networkNode.record, timing));
+    }
+  }
+
+  // Create a fake evaluate script event ~1s after the trace ends for a sane default bounds in DT
+  traceEvents.push(
+    createFakeTaskEvent(
+      {childEvents: []},
+      {
+        startTime: lastEventEndTime + 1000,
+        endTime: lastEventEndTime + 1001,
+      }
+    )
+  );
+
+  return {traceEvents};
+
+  /**
+   * @return {LH.TraceEvent}
+   */
+  function createFakeTracingStartedEvent() {
+    const argsData = {
+      frameTreeNodeId: 1,
+      sessionId: '1.1',
+      page: frame,
+      persistentIds: true,
+      frames: [{frame, url: 'about:blank', name: '', processId: 1}],
+    };
+
+    return {
+      ...baseEvent,
+      ts: baseTs - 1e5,
+      ph: 'I',
+      s: 't',
+      cat: 'disabled-by-default-devtools.timeline',
+      name: 'TracingStartedInPage',
+      args: {data: argsData},
+      dur: 0,
+    };
+  }
+
+  /**
+   * @param {{childEvents: LH.TraceEvent[]}} cpuNode
+   * @param {{startTime: number, endTime: number}} timing
+   * @return {LH.TraceEvent}
+   */
+  function createFakeTaskEvent(cpuNode, timing) {
+    const realEvaluateScriptEvent = cpuNode.childEvents.find(
+      e => e.name === 'EvaluateScript' && !!e.args.data && !!e.args.data.url
+    );
+    // @ts-ignore
+    const scriptUrl = realEvaluateScriptEvent && realEvaluateScriptEvent.args.data.url;
+    const argsData = {
+      url: scriptUrl || '',
+      frame,
+      lineNumber: 0,
+      columnNumber: 0,
+    };
+
+    return {
+      ...baseEvent,
+      ph: 'X',
+      name: 'Task',
+      ts: toMicroseconds(timing.startTime),
+      dur: (timing.endTime - timing.startTime) * 1000,
+      args: {data: argsData},
+    };
+  }
+
+  /**
+   * @param {LH.WebInspector.NetworkRequest} record
+   * @param {LH.Gatherer.Simulation.NodeTiming} timing
+   * @return {LH.TraceEvent[]}
+   */
+  function createFakeNetworkEvents(record, timing) {
+    requestId++;
+
+    // 0ms requests get super-messed up rendering
+    // Use 20ms instead so they're still hoverable
+    let {startTime, endTime} = timing; // eslint-disable-line prefer-const
+    if (startTime === endTime) endTime += 20;
```

Comment:
I still feel uncomfortable changing the data purely for flamechart UX reasons. E.g. a network request that _must_ start after a previous one finishes won't now. Feels weird, man.

How about... We do know there's pretty high overhead to create a network request. This is something we've discussed, but simulator doesn't acknowledge just yet. Let's add it and say request creation costs ~0.3ms or thereabouts.

Reply:
Alright 0.3ms it is 👍